### PR TITLE
Multi project classification

### LINKED ISSUE #101: Lineage Classification Averaging Multiple Embryos
# User Story

I would like to be able to process the lineage classification averaging over multiple embryos of the same species.
With this I would like to reproduced the steps describe in [Guignard et al. 2020, p. 46](https://www.science.org/doi/suppl/10.1126/science.aar5663/suppl_file/aar5663_guignard_sm.pdf)

grafik

grafik

# Acceptance criteria

1. Possibility for the user to select multiple Mastodon projects to be included in the analysis
2. Filter commonly named root cells across all selected projects
3. Crop all projects at the same given number of cells (start and end)
4. Compute Distance matrix for all projects
5. Compute an averaged distance matrix
6. Use the averaged distance matrix to compute an average dendromgram/classification
7. Save the resulting tag set in all projects included in the classification.

# Tasks

* [x] Possibility for the user to select multiple Mastodon projects to be included in the analysis
  * [x] ~Put Similarity measure, Linking method and feature into advanced options~
  * [x] ~Alternatively provide a dialog with only a view options and an advanced dialog with all params~
  * [x] Use Array of files as Sci Java Command param
  * [x] Clear button for list of projects to be included
* [x] Filter commonly named root cells across all selected projects
* [x] Crop all projects at the same given number of cells (start and end)
* [x] Compute Distance matrix for all projects
* [x] Compute an averaged distance matrix
* [x] Use the averaged distance matrix to compute an average dendrogram/classification
* [x] Handle edge cases / provide user feedback
* [x] (Optionally) Save the resulting tag set in all projects included in the classification.
  * [x] Name the tag set as "Average ...
* [x] Related PR: https://github.com/scijava/scijava-ui-swing/pull/82
* [x] Related PR: https://github.com/mastodon-sc/mastodon/pull/311
* [x] Related PR: https://github.com/scijava/scijava-common/pull/479
* [x] Mention name of project from which the plugin was loaded before the list of further projects
* [x] Avoid loading the current open project in the list of further projects 
* [x] Avoid that projects are loaded twice in the list of further projects

Optional

* normalize time across embryos
  * select one representative and normalize the time to this (embryo pm08 in the case of guignard et al. 2020)
  * normalize time using local cost function normalization (corrected guignard distance)

* evaluation of classification quality:
  * use concept of entropy to measure if the quality of the automatic classification in comparison to the classification done manually by Mette
  * https://towardsdatascience.com/statistical-tests-for-comparing-classification-algorithms-ac1804e79bb7

### PR DESCRIPTION
**Pull request by stefanhahmann**
Resolves #101

### CODE REVIEW

File: src/main/java/org/mastodon/mamut/classification/ClassifyLineagesController.java [hunk 1]
```diff
@@ -291,6 +419,73 @@ public void setVisualisationParams( final boolean showDendrogram )
 		this.showDendrogram = showDendrogram;
 	}
 
+	public void setExternalProjects( final File[] projects, final boolean addTagSetToExternalProjects )
```
Comment (maarzt) [inline src/main/java/org/mastodon/mamut/classification/ClassifyLineagesController.java:422]: I would recommend remove the ProjectSession, MastodonProjectService classes. They don't bring any advantage for the code. Instead I would recommend to add a class ExternalProjects which has this setExternalProjects(File[]) method, a close() method and probably some getter methods....

**Review by maarzt — APPROVED** [review]
Awesome test coverage! Also very well implemented user feedback in the Classify Lineages dialog.

The PR adds new functionality to allow lineage classification for multiple mastodon projects. The new functionality is not trivial. Unfortunately this also leads to growing classes and methods. Which makes the PR hard to review, and very like the code harder to maintain in the future. If possible please try to keep classes and methods small.

File: src/test/java/org/mastodon/mamut/classification/ClassifyLineagesControllerTest.java [hunk 2]
```diff
@@ -80,19 +87,19 @@ void testCreateTagSet()
...
+			addLineageTree51( modelGraph );
 			addEmptyTree( modelGraph );
 
 			String tagSetName = "Test Tag Set";
 			TagSetUtils.addNewTagSetToModel( model, tagSetName, Collections.emptyList() );
 			ClassifyLineagesController controller = new ClassifyLineagesController( projectModel );
 			controller.setInputParams( CropCriteria.TIMEPOINT, 0, 100, 1 );
 			controller.setComputeParams( SimilarityMeasure.NORMALIZED_ZHANG_DIFFERENCE, ClusteringMethod.AVERAGE_LINKAGE, 3 );
-			controller.setVisualisationParams( false );
+			controller.setVisualisationParams( true );
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/test/java/org/mastodon/mamut/classification/ClassifyLineagesControllerTest.java:102]: Maybe consider renaming the method to setShowDendrogram.

File: src/main/java/org/mastodon/mamut/classification/ClassifyLineagesController.java [hunk 3]
```diff
@@ -90,64 +106,167 @@ public class ClassifyLineagesController
...
+						try
+						{
+							ProjectSaver.saveProject( file, projectModel );
+						}
+						catch ( IOException e )
+						{
+							logger.warn( "Could not save tag set of project {} to file {}. Message: {}", projectModel.getProjectName(),
+									file.getAbsolutePath(), e.getMessage() );
+						}
+					}
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/classification/ClassifyLineagesController.java]: Good idea. Done.

File: src/main/java/org/mastodon/mamut/classification/util/Classification.java [hunk 4]
```diff
@@ -57,7 +57,7 @@
  */
 public class Classification< T >
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/main/java/org/mastodon/mamut/classification/util/Classification.java:58]: I added hierarchical to the javadoc of the class and the fact that the contained cluster are hierarchically organized.

File: src/test/java/org/mastodon/mamut/classification/ClassifyLineagesControllerTest.java [hunk 5]
```diff
@@ -121,6 +128,124 @@ void testCreateTagSet()
...
+			controller.setExternalProjects( files, true );
+			controller.createTagSet();
+
+			Set< String > expectedClassNames = new HashSet<>( Arrays.asList( "Class 1", "Class 2", "Class 3" ) );
+			Set< Integer > expectedClassCounts = new HashSet<>( Arrays.asList( 9, 12, 14 ) );
+
+			assertTrue( controller.isValidParams() );
+			assertClassificationEquals( model1, 1, expectedClassNames, expectedClassCounts );
+			ProjectModel pm2 = ProjectLoader.open( file2.getAbsolutePath(), context, false, true );
+			assertClassificationEquals( pm2.getModel(), 0, expectedClassNames, expectedClassCounts );
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply src/test/java/org/mastodon/mamut/classification/ClassifyLineagesControllerTest.java]: That's actually a good idea. Having the possibility to load and visualize the test datasets in Mastodon makes the assumptions tested here more transparent.

I changed it in this direction.

**Pull request merged**